lib/mediastream.js

```diff
@@ -0,0 +1,66 @@
+var EventTarget = require('./eventtarget');
+
+var RTCMediaStreamMessageEvent = require('./mediastreammessageevent');
+
+function RTCMediaStream(internalMS) {
+  'use strict';
+  var that = this;
+
+  EventTarget.call(this);
+
+  that._internalMS = internalMS;
+  internalMS.onerror = function onerror() {
+    that.dispatchEvent({type: 'error'});
+  };
+
+  internalMS.onmessage = function onmessage(data) {
+    var evt = new RTCMediaStreamMessageEvent(data);
+    that.dispatchEvent(evt);
+    that._onmessage && that._onmessage(evt);
+  };
+
+  that.getAudioTracks = function() {
+    return internalMS.getAudioTracks();
+  }
+  that.getVideoTracks = function() {
+    return internalMS.getVideoTracks();
+  }
+  that.getTracks = function() {
+    return internalMS.getVideoTracks().concat(internalMS.getAudioTracks());
+  }
+  Object.defineProperties(this, {
+    'id': {
+        get: function getID() {
+            return internalMS.id;
+        }
+    },
+    'active': {
+        get: function() {
+          return internalMS.active;
+        }
+    },
+    'ended': {
+        get: function() {
+          return internalMS.readyState === 'closed';
+        }
+    },
+    'onmessage': {
+        set: function(onMessage) {
+            that._onmessage = onMessage;
+        }
+    }
+  });
+
+  this.close = function close() {
+    internalMS.close();
+  };
+}
+
+RTCMediaStream.prototype.RTCMediaStates = [
+  'connecting',
+  'open',
+  'closing',
+  'closed'
+];
+
+module.exports = RTCMediaStream;
```

Comment:
Nice. I have a sort of mockup of MediaStreamTrack in my branch that just composes an object from Javascript, but it definitely seems better to avoid the wrapper.

Just a question--- does this make it so in javascript land a track would pass type checks against RTCMediaStreamTrack?

There is some nicety in being able to wrap and modify the interface to these things in Javascript rather than in C++, particularly if in the future we wanted to more easily provide support for spec changes without having to rewrite a significant portion of the C++ interface.

Reply:
> Just a question--- does this make it so in javascript land a track would pass type checks against RTCMediaStreamTrack?

Sorry, I'm not sure what you mean. Can you ask in a different way? With either approach, given some `mediaStreamTrack` and `mediaStream`, we can satisfy

```js
const { MediaStream, MediaStreamTrack } = require('wrtc')
mediaStream instanceof MediaStream
mediaStreamTrack instanceof MediaStreamTrack
```

if we want to. We can also ensure methods that should only accept proper instances of these accept only proper instances, too. There's just more wrapping and unwrapping necessary if we have JavaScript wrappers.

> There is some nicety in being able to wrap and modify the interface to these things in Javascript rather than in C++, particularly if in the future we wanted to more easily provide support for spec changes without having to rewrite a significant portion of the C++ interface.

Yes, this is true. For example, it might be nice to shim the legacy APIs in JavaScript rather than bind to the native APIs (I'm thinking `addStream`, or the callback versions of `setLocalDescription`, etc.—unless the shims are complicated enough and libwebrtc actually provides correct legacy APIs). As far as newly specced APIs, I'm not sure what's easier in JavaScript right now. But we _can_ still shim methods C++ classes' `prototype` if necessary.